sdk/python/foundation-models/system/inference/image-text-embeddings/text-to-image-retreival.ipynb

```diff
@@ -0,0 +1,586 @@
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from tqdm.auto import tqdm\n",
+    "image_paths = [os.path.join(dp, f) for dp, dn, filenames in os.walk(dataset_dir) for f in filenames if os.path.splitext(f)[1] == '.jpg']\n",
+    "\n",
+    "for idx, image_path in enumerate(tqdm(image_paths)):\n",
+    "    ID = idx\n",
+    "    FILENAME = image_path\n",
+    "    MAX_RETRIES = 3\n",
+    "    \n",
+    "    # get embedding from endpoint\n",
+    "    embedding_request = make_request_images(image_path)\n",
+    "\n",
+    "    response = None\n",
+    "    request_failed = False\n",
+    "    IMAGE_EMBEDDING = None\n",
+    "    for r in range(MAX_RETRIES):\n",
+    "        try:\n",
+    "            response = workspace_ml_client.online_endpoints.invoke(\n",
+    "                endpoint_name=online_endpoint_name,\n",
+    "                deployment_name=deployment_name,\n",
+    "                request_file=_REQUEST_FILE_NAME,\n",
+    "            )\n",
+    "            response = json.loads(response)\n",
+    "            IMAGE_EMBEDDING = response[0]['image_features']\n",
+    "            break\n",
+    "        except Exception as e:\n",
+    "            print(f\"Unable to get embeddings for image {FILENAME}: {e}\")\n",
+    "            print(response)\n",
+    "            if r == MAX_RETRIES - 1:\n",
+    "                print(f\"attempt {r} failed, reached retry limit\")\n",
+    "                request_failed = True\n",
+    "            else:\n",
+    "                print(f\"attempt {r} failed, retrying\")\n",
+    "\n",
+    "    # add embedding to index\n",
+    "    if IMAGE_EMBEDDING:\n",
+    "        add_data_request = {\n",
+    "            \"value\": [\n",
+    "                {\n",
+    "                    \"id\": str(ID),\n",
+    "                    \"filename\": FILENAME,\n",
+    "                    \"imageEmbeddings\": IMAGE_EMBEDDING,\n",
+    "                    \"@search.action\": \"upload\"\n",
+    "                }\n",
+    "            ]\n",
+    "        }\n",
+    "        response = requests.post(ADD_DATA_REQUEST_URL, json=add_data_request, headers={\"api-key\": SERVICE_ADMIN_KEY})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 6. Query the index with text embeddings and visualize results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "TEXT_QUERY = \"a photo of a milk bottle\"\n",
+    "K = 10 # number of results to retrieve"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 6.1 Get the text embeddings for the query using the online endpoint"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def make_request_text(text_sample):\n",
+    "    request_json = {\n",
```

Comment:
here I see that image part is left to be empty. I know it's a text-to-image retrieval, but is it also true that it could do image to image retrieval? since for clip model it was able to return image imbedding as well right? does image embedding & text embedding has same shape? - was just curious.

Reply:
Yes we can do image to image retrieval! Simply we would send a query image to the endpoint and then send the returned embeddings to the search index. The image and text embeddings from clip model do have the same shape which is why they are comparable (multi-modal embeddings)